examples/custom-xviz-layers/src/app.js

```diff
@@ -0,0 +1,172 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* global document, console */
+/* eslint-disable no-console, no-unused-vars, no-undef */
+import React, {PureComponent} from 'react';
+import {render} from 'react-dom';
+import {PathStyleExtension} from '@deck.gl/extensions';
+import {setXVIZConfig, getXVIZConfig} from '@xviz/parser';
+import {
+  LaneLayer,
+  LogViewer,
+  PlaybackControl,
+  StreamSettingsPanel,
+  MeterWidget,
+  TrafficLightWidget,
+  TurnSignalWidget,
+  XVIZPanel,
+  VIEW_MODE
+} from 'streetscape.gl';
+import {Form} from '@streetscape.gl/monochrome';
+
+import {XVIZ_CONFIG, APP_SETTINGS, MAPBOX_TOKEN, MAP_STYLE, XVIZ_STYLE, CAR} from './constants';
+
+setXVIZConfig(XVIZ_CONFIG);
+
+const TIMEFORMAT_SCALE = getXVIZConfig().TIMESTAMP_FORMAT === 'seconds' ? 1000 : 1;
+
+// __IS_STREAMING__ and __IS_LIVE__ are defined in webpack.config.js
+const exampleLog = require(__IS_STREAMING__
+  ? './log-from-stream'
+  : __IS_LIVE__
+    ? './log-from-live'
+    : './log-from-file').default;
+
+class Example extends PureComponent {
+  state = {
+    log: exampleLog,
+    settings: {
+      viewMode: 'PERSPECTIVE',
+      showTooltip: false
+    }
+  };
+
+  componentDidMount() {
+    this.state.log.on('error', console.error).connect();
+  }
+
+  _onSettingsChange = changedSettings => {
+    this.setState({
+      settings: {...this.state.settings, ...changedSettings}
+    });
+  };
+
+  render() {
+    const {log, settings} = this.state;
+
+    return (
+      <div id="container">
+        <div id="control-panel">
+          <XVIZPanel log={log} name="Metrics" />
+          <hr />
+          <XVIZPanel log={log} name="Camera" />
+          <hr />
+          <Form
+            data={APP_SETTINGS}
+            values={this.state.settings}
+            onChange={this._onSettingsChange}
+          />
+          <StreamSettingsPanel log={log} />
+        </div>
+        <div id="log-panel">
+          <div id="map-view">
+            <LogViewer
+              log={log}
+              mapboxApiAccessToken={MAPBOX_TOKEN}
+              mapStyle={MAP_STYLE}
+              car={CAR}
+              xvizStyles={XVIZ_STYLE}
+              showTooltip={settings.showTooltip}
+              viewMode={VIEW_MODE[settings.viewMode]}
+              customXVIZLayers={[
+                {
+                  // {string} id - passed into layer
+                  id: 'dual',
+                  //
+                  extendPrimitive: 'path',
```

Comment:
This is referring to the Deck.gl layer used?  XVIZ only defines the type 'polyline'.  So this seems to expose an implementation detail on what deck layer is used.  Is this what we want to do?

I guess this is really only for streetscape.gl, so probably ok, but i' not sure if we exposed this anywhere else.

Reply:
not quite, it corresponds to the primitive that links to the layer within the XVIZLayer